DOC Add minimal reproducer guide for scikit-learn (#22401)
scikit-learn/scikit-learn: doc/developers/minimal_reproducer.rst

<https://github.com/scikit-learn/scikit-learn/blob/main/.github/ISSUE_TEMPLATE/bug_report.yml>`_.
We will start with a snippet of code that already provides a failing example but
that has room for readability improvement. We then craft a MCVE from it.

**Example**

.. code-block:: python

    # I am currently working in a ML project and when I tried to fit a
    # GradientBoostingRegressor instance to my_data.csv I get a UserWarning:
    # "X has feature names, but DecisionTreeRegressor was fitted without feature
    # names". You can get a copy of my dataset from
    # https://example.com/my_data.csv and verify my features do have
    # names. The problem seems to arise during fit when I pass an integer
    # to the n_iter_no_change parameter.

    df = pd.read_csv('my_data.csv')
    X = df[["feature_name"]] # my features do have names
    y = df["target"]

    # We set random_state=42 for the train_test_split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=42)

    # An instance with default n_iter_no_change raises no error nor warnings
    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)

> 

Reviewer: ```suggestion

    default_score = gbdt.score(X_test, y_test)

```

    # the bug appears when I change the value for n_iter_no_change
    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)

>     gbdt.fit(X_train, y_train)

Reviewer: ```suggestion
    gbdt.fit(X_train, y_train)
    other_score = gbdt.score(X_test, y_test)

```

Provide a failing code example with minimal comments
----------------------------------------------------

> Instructions may get lost in translation for non-native english speakers.

Reviewer: ```suggestion
Instructions may get lost in translation for non-native English speakers.
```

    y = df["target"]

    # We set random_state=42 for the train_test_split
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=42)

    from sklearn.ensemble import GradientBoostingRegressor

    # An instance with default n_iter_no_change raises no error nor warnings
    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)

    # the bug appears when I change the value for n_iter_no_change
    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
    gbdt.fit(X_train, y_train)

Provide a failing code example with minimal annotations
-------------------------------------------------------

Please take into account that things may get lost in translation for non-native
english speakers. It is better that all the necessary information can be read
from the code itself. Besides, by this point you already provided a concise
description in the **Describe the bug** section of the Issue template.

**Better example**

.. code-block:: python

    import pandas as pd
    df = pd.read_csv('my_data.csv')

>     X = df["feature_name"]

Author: ```suggestion
    X = df[["feature_name"]]
```

    y = df["target"]

    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=42)

    from sklearn.ensemble import GradientBoostingRegressor

    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train) # compiles

    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
    gbdt.fit(X_train, y_train) # raises warning

Boil down your script to something as small as possible
-------------------------------------------------------

You have to ask yourself which lines of code are relavant and which are not for
rasing the bug. Erasing unnecessary lines of code will help you and other
contributors narrow down the bug. In this example the warning has nothing to
do with the `train_test_split`, nor the `random_state`.

**Even better example**

.. code-block:: python

    import pandas as pd
    df = pd.read_csv('my_data.csv')

>     X = df["feature_name"]

Author: ```suggestion
    X = df[["feature_name"]]
```

    y = df["target"]

    from sklearn.ensemble import GradientBoostingRegressor

    gbdt = GradientBoostingRegressor()
    gbdt.fit(X, y) # compiles

    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
    gbdt.fit(X, y) # raises warning

**DO NOT** report your data unless it is extremely necessary
------------------------------------------------------------

The idea is to make the code as self-contained as possible. For doing so, you
can use a :ref:`synth_data`. It can be generated using numpy, pandas or the
:mod:`sklearn.datasets` module. Most of the times the bug is not related to a
particular structure of your data. Even if it is, try to find an available
dataset that has similar characteristics to yours and that reproduces the
problem. In this particular case, we are interested in data that has labeled
feature names.

**Even better example**

.. code-block:: python

    import pandas as pd
    df = pd.DataFrame(
        {
            "feature_name": [-12.32, 1.43, 30.00],
            "target": [72, 55, 32],
        }
    )

>     X = df["feature_name"]

Author: ```suggestion
    X = df[["feature_name"]]
```

    gbdt.fit(X, y) # compiles

    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
    gbdt.fit(X, y) # raises warning

Improve your formatting as much as possible
-------------------------------------------

As already mentioned, the key to communication is the readability of the code
and good formatting can really improve it.

Try to follow the `PEP 8 convention <https://www.python.org/dev/peps/pep-0008/>`_.
In a nutshell:

>     - try to limit all lines to a maximum of 79 characters

Author: Thanks @cmarmo !

    ---------------------------------------------------------------------------
    TypeError                                 Traceback (most recent call last)
    <ipython-input-1-a674e682c281> in <module>
        4 vectorizer = CountVectorizer(input=docs, analyzer='word')
        5 lda_features = vectorizer.fit_transform(docs)
    ----> 6 lda_model = LatentDirichletAllocation(
        7     n_topics=10,
        8     learning_method='online',

    TypeError: __init__() got an unexpected keyword argument 'n_topics'

Try to follow the `PEP 8 convention <https://www.python.org/dev/peps/pep-0008/>`_
---------------------------------------------------------------------------------

The convention in a nutshell:
    - try to limit all lines to a maximum of 79 characters
    - use blank lines to separate groups of related functions
    - blank lines may be omitted between a bunch of related lines of code

**Example**

The MCVE we created in the :ref:`good_practices` section is easier to read than
the equally working MCVE here below

.. code-block:: python

    import pandas as pd
    df = pd.DataFrame({"feature_name": [-12.32, 1.43, 30.00], "target": [72, 55, 32]})

>     X = df["feature_name"]

Author: ```suggestion
    X = df[["feature_name"]]
```

step-by-step guide on how to narrow down a bug until you have reached the
shortest possible code to reproduce it.

The first step before submitting a bug report to scikit-learn is to read the
`Issue template
<https://github.com/scikit-learn/scikit-learn/blob/main/.github/ISSUE_TEMPLATE/bug_report.yml>`_.
It is already quite informative about the information you will be asked to provide.

.. _good_practices:

Good practices
==============

In this section we will focus on the **Steps/Code to Reproduce** section of
the Issue template. We will start with a snippet of code that already provides a
failing example but that has room for readability improvement. We then craft a
MCVE from it.

**Example**

.. code-block:: python

    # I am currently working in a ML project and when I tried to fit a
    # GradientBoostingRegressor instance to my_data.csv I get a UserWarning:
    # "X has feature names, but DecisionTreeRegressor was fitted without feature
    # names". You can get a copy of my dataset from www.this_link.com and verify
    # my features do have names. The problem seems to arise during fit when I pass
    # an integer to the n_iter_no_change parameter.

>     import pandas as pd

Reviewer: A common error is to skip the imports. Let's forget them in the first  (bad) snippet.

```suggestion
```

    df = pd.read_csv('my_data.csv')
    X = df[["feature_name"]] # my features do have names
    y = df["target"]

    # We set random_state=42 for the train_test_split
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=42)

    from sklearn.ensemble import GradientBoostingRegressor

    # An instance with default n_iter_no_change raises no error nor warnings
    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)

    # the bug appears when I change the value for n_iter_no_change
    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
    gbdt.fit(X_train, y_train)

Provide a failing code example with minimal comments
----------------------------------------------------

Instructions may get lost in translation for non-native english speakers.
Hence, it is better that all the necessary information can be read
from the code itself. Besides, by this point you already provided a concise
description in the **Describe the bug** section of the Issue template.

**Improved example**

> 

Reviewer: ```suggestion
The following code, while **still not minimal**, is already **much better** because
it can be copy-pasted in a python terminal to reproduce the problem in one step.
In particular:
- it contains **all necessary imports statements**;
- it can fetch the public dataset without having to manually download a
  file and put it in the expected location on the disk.
```

    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
    gbdt.fit(X, y) # raises warning

**DO NOT** report your data unless it is extremely necessary
------------------------------------------------------------

The idea is to make the code as self-contained as possible. For doing so, you
can use a :ref:`synth_data`. It can be generated using numpy, pandas or the
:mod:`sklearn.datasets` module. Most of the times the bug is not related to a
particular structure of your data. Even if it is, try to find an available
dataset that has similar characteristics to yours and that reproduces the
problem. In this particular case, we are interested in data that has labeled
feature names.

**Improved example**

.. code-block:: python

    import pandas as pd
    df = pd.DataFrame({"feature_name": [-12.32, 1.43, 30.01, 22.17], "target": [72, 55, 32, 43]})
    X = df[["feature_name"]]
    y = df["target"]

    from sklearn.ensemble import GradientBoostingRegressor

    gbdt = GradientBoostingRegressor()
    gbdt.fit(X, y) # compiles

    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
    gbdt.fit(X, y) # raises warning

Improve your formatting as much as possible

> -------------------------------------------

Reviewer: I am not so sure this step is necessary.

While following code conventions is important when contributing a PR, it's not as important when crafting a minimal reproducer. I would therefore shorten this tutorial by removing this step and not mentioning PEP8 and other formatting concerns.

Instead we could directly improve the formatting of the previous step to directly show a good example of well formatted minimal reproducer implicitly.

Author: I tried to keep the guide relatively general for concordance with the first paragraph:

> Whether submitting a bug report, designing a suite of tests, or simply posting a
question in the discussions, being able to craft minimal, reproducible examples
(or minimal, workable examples) is the key to communicating effectively and
efficiently with the community.

I don't have strong feelings about keeping this step, but if you think it is making this guide too long, maybe we can send this part to the end of the page as an "extra"/"nice-to-have", what do you think?

Reviewer: I think it's better to remove the section on formatting and use a better code style from the beginning (show by example) and focus the content of this guide on the most important points (simplifying the code and test data, including the imports).

Let's try not be too nitpicky for bug reports.

As already mentioned, the key to communication is the readability of the code
and good formatting can really improve it.

Try to follow the `PEP 8 convention <https://www.python.org/dev/peps/pep-0008/>`_.
In a nutshell:

    - try to limit all lines to a maximum of 79 characters
    - use blank lines to separate groups of related functions

Place all your imports in their own group at the beginning of the snippet.

**Improved example**

.. code-block:: python

    import pandas as pd
    from sklearn.ensemble import GradientBoostingRegressor

    df = pd.DataFrame(
        {
            "feature_name": [-12.32, 1.43, 30.01, 22.17],
            "target": [72, 55, 32, 43],
        }
    )
    X = df[["feature_name"]]
    y = df["target"]

    gbdt = GradientBoostingRegressor()
    gbdt.fit(X, y) # compiles
    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
    gbdt.fit(X, y) # raises warning

> Use markdown
> ------------

Reviewer: ```suggestion
Use markdown formatting
-----------------------
```

    TypeError                                 Traceback (most recent call last)
    <ipython-input-1-a674e682c281> in <module>
        4 vectorizer = CountVectorizer(input=docs, analyzer='word')
        5 lda_features = vectorizer.fit_transform(docs)
    ----> 6 lda_model = LatentDirichletAllocation(
        7     n_topics=10,
        8     learning_method='online',

    TypeError: __init__() got an unexpected keyword argument 'n_topics'
    ```

yields the following when rendered:

.. code-block:: python

    ---------------------------------------------------------------------------
    TypeError                                 Traceback (most recent call last)
    <ipython-input-1-a674e682c281> in <module>
        4 vectorizer = CountVectorizer(input=docs, analyzer='word')
        5 lda_features = vectorizer.fit_transform(docs)
    ----> 6 lda_model = LatentDirichletAllocation(
        7     n_topics=10,
        8     learning_method='online',

    TypeError: __init__() got an unexpected keyword argument 'n_topics'

The above steps can be implemented in a different order than the progression we
show in this example. Take into account that having a pipeline that makes sense

> as a model is not important when creating a MCVE.

Reviewer: I do not understand this sentence. Maybe we can just remove it.

If we want to illustrate that complex preprocessing can be useless to reproduce a problem, I would instead add unnecessary steps to the first code snippets and then add a new section to explain that it's important to try to simplify the model structure to remove any step (pipeline, columntransformers, ...) that are not needed to reproduce the problem.

Reviewer: I still don't get it. Maybe the

"""
The simplification steps presented in this guide can be implemented in a different order than the progression we have shown here. The important points are:

- a minimal reproducer should be runnable by a simple copy and paste in a python terminal;
- it should be simplified as much as possible by removing any code steps that are not strictly needed to reproducing the original problem;
- if should ideally only rely on a minimal dataset generated on-the-fly by running the code instead of relying on external data, if possible.
"""

Reviewer: But this is not related to markdown formatting. So it would better be moved before the section on markdown formatting.

.. _synth_data:

Synthetic dataset
=================

> Before chosing a particular synthetic dataset, first you have to identify the
> type of problem you are solving: Is it classification, regression, clustering,
> etc?

Reviewer 2: ```suggestion
Before choosing a particular synthetic dataset, first you have to identify the
type of problem you are solving: Is it a classification, a regression,
a clustering, etc?
```

Once that you narrowed down the type of problem, you need to provide a synthetic
dataset accordingly. Most of the times you only need a minimalistic dataset.
Here is a non-exhaustive list of tools that may help you.

> Numpy

Reviewer: ```suggestion
NumPy
```

-----

> Numpy tools such as `random.randn

Reviewer: ```suggestion
NumPy tools such as `numpy.random.randn
```

<https://numpy.org/doc/stable/reference/random/generated/numpy.random.randn.html>`_

> and `random.randint

Reviewer: ```suggestion
and `numpy.random.randint
```

to avoid confusion with the `random` module of the standard library.

<https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html>`_
can be used to create dummy numeric data.

- regression

    Regressions take continuous numeric data as features and target.

    .. code-block:: python

        import numpy as np

        rng = np.random.RandomState(0)
        n_samples, n_features = 5, 5
        X = rng.randn(n_samples, n_features)
        y = rng.randn(n_samples)

A similar snippet can be used as synthetic data when testing scaling tools such
as :class:`sklearn.preprocessing.StandardScaler`.

- classification

>     If the bug is not raised during encoding a categorical variable, you can
>     feed numeric data to a classifier. Just remember to ensure that the target
>     is indeed an integer.

Reviewer 2: ```suggestion
    If the bug is not raised during when encoding a categorical variable, you can
    feed numeric data to a classifier. Just remember to ensure that the target
    is indeed an integer.
```

    .. code-block:: python

        import numpy as np

        rng = np.random.RandomState(0)
        n_samples, n_features = 5, 5
        X = rng.randn(n_samples, n_features)

>         y = rng.randint(0, 2, n_samples)

Reviewer: ```suggestion
        y = rng.randint(0, 2, n_samples)  # binary target with values in {0, 1}
```

    If you need to test encoding, you may prefer to start from non-numeric data.
    In such case you may use `numpy.random.choice
    <https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html>`_.

    .. code-block:: python

        import numpy as np

        rng = np.random.RandomState(0)
        n_samples, n_features = 50, 5
        X = rng.randn(n_samples, n_features)
        y = np.random.choice(
            ["male", "female", "other"], size=n_samples, p=[0.49, 0.49, 0.02]
        )

Pandas
------

Some scikit-learn objets expect pandas dataframes as input. In this case you can
transform numpy arrays into pandas objects using `pandas.DataFrame
<https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`_, or
`pandas.Series
<https://pandas.pydata.org/docs/reference/api/pandas.Series.html>`_.

.. code-block:: python

    import numpy as np
    import pandas as pd

    rng = np.random.RandomState(0)
    n_samples, n_features = 5, 5

>     X = rng.randn(n_samples, n_features)
>     y = rng.randn(n_samples)
>     X = pd.DataFrame(X)
>     y = pd.Series(y)

Reviewer: Make the example dataframe more interesting:

```suggestion
    X = pd.DataFrame(
        {
            "continuous_feature": rng.randn(n_samples),
            "positive_feature": rng.uniform(low=0.0, high=100.0, size=n_samples),
            "categorical_feature": rng.choice(["a", "b", "c"], size=n_samples),
        }
    )
    y = pd.Series(rng.randn(n_samples))
```

clusters per class. Noise can be introduced by means of correlated, redundant or
uninformative features.

.. code-block:: python

    from sklearn.datasets import make_classification

    X, y = make_classification(
        n_features=2, n_redundant=0, n_informative=2, n_clusters_per_class=1
    )

`make_blobs`
------------

Similarly to `make_classification`, :class:`sklearn.datasets.make_blobs` creates multiclass
datasets using normally-distributed clusters of points. It provides greater
control regarding the centers and standard deviations of each cluster, and
therefore it is useful to demonstrate clustering.

.. code-block:: python

    from sklearn.datasets import make_blobs

    X, y = make_blobs(n_samples=10, centers=3, n_features=2)

Dataset loading utilities
-------------------------

You can use the :ref:`datasets` to load and fetch several popular reference
datasets. This option is useful when the bug relates to the particular structure
of the data, e.g. dealing with missing values or image recognition.

.. code-block:: python

    from sklearn.datasets import load_breast_cancer

>     cancer = load_breast_cancer()
>     X = cancer.data
>     y = cancer.target

Reviewer: Conciseness is useful for minimal reproducers :)

```suggestion
    X, y = load_breast_cancer(return_X_y=True)
```

It is already quite informative about the information you will be asked to provide.

.. _good_practices:

Good practices
==============

In this section we will focus on the **Steps/Code to Reproduce** section of
the Issue template. We will start with a snippet of code that already provides a
failing example but that has room for readability improvement. We then craft a
MCVE from it.

**Example**

.. code-block:: python

    # I am currently working in a ML project and when I tried to fit a
    # GradientBoostingRegressor instance to my_data.csv I get a UserWarning:
    # "X has feature names, but DecisionTreeRegressor was fitted without feature
    # names". You can get a copy of my dataset from
    # https://example.com/my_data.csv and verify my features do have
    # names. The problem seems to arise during fit when I pass an integer
    # to the n_iter_no_change parameter.

    df = pd.read_csv('my_data.csv')
    X = df[["feature_name"]] # my features do have names
    y = df["target"]

    # We set random_state=42 for the train_test_split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=42)

> 

Reviewer: Let's add some more unnecessary code similar to what we typically observe in non-minimal bug reports to this code snippet:

```suggestion

    scaler = StandardScaler(with_mean=False)
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)
```

The same lines wil

    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)

    # the bug appears when I change the value for n_iter_no_change
    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
    gbdt.fit(X_train, y_train)

Provide a failing code example with minimal comments
----------------------------------------------------

Instructions may get lost in translation for non-native english speakers.
Hence, it is better that all the necessary information can be read
from the code itself. Besides, by this point you already provided a concise
description in the **Describe the bug** section of the Issue template.

**Improved example**
The following code, while **still not minimal**, is already **much better** because
it can be copy-pasted in a python terminal to reproduce the problem in one step.
In particular:
- it contains **all necessary imports statements**;
- it can fetch the public dataset without having to manually download a
  file and put it in the expected location on the disk.
.. code-block:: python

    import pandas as pd
    df = pd.read_csv("https://example.com/my_data.csv")
    X = df[["feature_name"]]

>     y = df["target"]

Reviewer: ```suggestion
    y = df["target"]

    from sklearn.preprocessing import scale

    X = scale(X)
```

    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.33, random_state=42)

> 

Reviewer: ```suggestion

    from sklearn.preprocessing import StandardScaler

    scaler = StandardScaler(with_mean=False)
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)
```

    from sklearn.ensemble import GradientBoostingRegressor

    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)  # no warning

> 

Reviewer: More unrelated lines:

```suggestion
    default_score = gbdt.score(X_test, y_test)

```

    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
    gbdt.fit(X_train, y_train) # raises warning

> 

Reviewer: ```suggestion
    other_score = gbdt.score(X_test, y_test)

```

Boil down your script to something as small as possible
-------------------------------------------------------

> You have to ask yourself which lines of code are relevant and which are not for
> raising the bug. Erasing unnecessary lines of code will help you and other
> contributors narrow down the bug. In this example the warning has nothing to
> do with the `train_test_split`, nor the `random_state`.

Reviewer: ```suggestion
You have to ask yourself which lines of code are relevant and which are not for
reproducing the bug. Deleting unnecessary lines of code or simplifying the function
calls by omitting unrelated non-default options will help you and other contributors
narrow down the cause of the bug.

In particular, for  this specific example:

- the warning has nothing to do with the `train_test_split` since it already appears
   in the training set.
- similarly, the lines that compute the scores on the test set are not necessary;
- the bug can be reproduced for any value of`random_state` so leave it to its
   default;
- the bug can be reproduced without preprocessing the data with the
  `StandardScaler`.
```

.. _minimal_reproducer:

> ===========================================
> Craft a minimal reproducer for scikit-learn
> ===========================================

Reviewer: ```suggestion
=============================================
Crafting a minimal reproducer for scikit-learn
=============================================
```

provide.

.. _good_practices:

Good practices
==============

In this section we will focus on the **Steps/Code to Reproduce** section of the
`Issue template
<https://github.com/scikit-learn/scikit-learn/blob/main/.github/ISSUE_TEMPLATE/bug_report.yml>`_.
We will start with a snippet of code that already provides a failing example but
that has room for readability improvement. We then craft a MCVE from it.

**Example**

.. code-block:: python

    # I am currently working in a ML project and when I tried to fit a
    # GradientBoostingRegressor instance to my_data.csv I get a UserWarning:
    # "X has feature names, but DecisionTreeRegressor was fitted without
    # feature names". You can get a copy of my dataset from
    # https://example.com/my_data.csv and verify my features do have
    # names. The problem seems to arise during fit when I pass an integer
    # to the n_iter_no_change parameter.

    df = pd.read_csv('my_data.csv')
    X = df[["feature_name"]] # my features do have names
    y = df["target"]

    # We set random_state=42 for the train_test_split
    X_train, X_test, y_train, y_test = train_test_split(

>         X, y, test_size=0.33, random_state=42)

Reviewer: any color you like, ... as long as it's black:

```suggestion
        X, y, test_size=0.33, random_state=42
    )
```

    scaler = StandardScaler(with_mean=False)
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    # An instance with default n_iter_no_change raises no error nor warnings
    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)

> 

Reviewer: ```suggestion
```

    default_score = gbdt.score(X_test, y_test)

    # the bug appears when I change the value for n_iter_no_change
    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
    gbdt.fit(X_train, y_train)
    other_score = gbdt.score(X_test, y_test)

Provide a failing code example with minimal comments
----------------------------------------------------

Writing instructions to reproduce the problem in English is often ambiguous.
Better make sure that all the necessary details to reproduce the problem are
illustrated in the Python code snippet to avoid any ambiguity. Besides, by this
point you already provided a concise description in the **Describe the bug**
section of the `Issue template
<https://github.com/scikit-learn/scikit-learn/blob/main/.github/ISSUE_TEMPLATE/bug_report.yml>`_.

The following code, while **still not minimal**, is already **much better**

> because it can be copy-pasted in a python terminal to reproduce the problem in

Reviewer: Sorry @lesteve, this was my fault:

```suggestion
because it can be copy-pasted in a Python terminal to reproduce the problem in
```

one step. In particular:

    - it contains **all necessary imports statements**;
    - it can fetch the public dataset without having to manually download a
      file and put it in the expected location on the disk.

**Improved example**

.. code-block:: python

>     import pandas as pd

Reviewer: For consistency with other import statements within this code snippet.

```suggestion
    import pandas as pd

```

    df = pd.read_csv("https://example.com/my_data.csv")
    X = df[["feature_name"]]
    y = df["target"]

>     from sklearn.model_selection import train_test_split

Reviewer: For consistency with other import statements within this code snippet.

```suggestion
    from sklearn.model_selection import train_test_split

```

    X_train, X_test, y_train, y_test = train_test_split(

>         X, y, test_size=0.33, random_state=42)

Reviewer: ```suggestion
        X, y, test_size=0.33, random_state=42
    )
```

    from sklearn.preprocessing import StandardScaler

    scaler = StandardScaler(with_mean=False)
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    from sklearn.ensemble import GradientBoostingRegressor

    gbdt = GradientBoostingRegressor(random_state=0)
    gbdt.fit(X_train, y_train)  # no warning
    default_score = gbdt.score(X_test, y_test)

    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)

>     gbdt.fit(X_train, y_train) # raises warning

Reviewer: ```suggestion
    gbdt.fit(X_train, y_train)  # raises warning
```

-------------------------------------------------------

You have to ask yourself which lines of code are relevant and which are not for
reproducing the bug. Deleting unnecessary lines of code or simplifying the
function calls by omitting unrelated non-default options will help you and other
contributors narrow down the cause of the bug.

In particular, for this specific example:

- the warning has nothing to do with the `train_test_split` since it already
  appears in the training step, before we use the test set.
- similarly, the lines that compute the scores on the test set are not
  necessary;
- the bug can be reproduced for any value of `random_state` so leave it to its
  default;
- the bug can be reproduced without preprocessing the data with the
  `StandardScaler`.

**Improved example**

.. code-block:: python

    import pandas as pd
    df = pd.read_csv("https://example.com/my_data.csv")
    X = df[["feature_name"]]
    y = df["target"]

    from sklearn.ensemble import GradientBoostingRegressor

    gbdt = GradientBoostingRegressor()
    gbdt.fit(X, y)  # no warning

    gbdt = GradientBoostingRegressor(n_iter_no_change=5)

>     gbdt.fit(X, y) # raises warning

Reviewer: ```suggestion
    gbdt.fit(X, y)  # raises warning
```

------------------------------------------------------------

The idea is to make the code as self-contained as possible. For doing so, you
can use a :ref:`synth_data`. It can be generated using numpy, pandas or the
:mod:`sklearn.datasets` module. Most of the times the bug is not related to a
particular structure of your data. Even if it is, try to find an available
dataset that has similar characteristics to yours and that reproduces the
problem. In this particular case, we are interested in data that has labeled
feature names.

**Improved example**

.. code-block:: python

    import pandas as pd
    from sklearn.ensemble import GradientBoostingRegressor

    df = pd.DataFrame(
        {
            "feature_name": [-12.32, 1.43, 30.01, 22.17],
            "target": [72, 55, 32, 43],
        }
    )
    X = df[["feature_name"]]
    y = df["target"]

    gbdt = GradientBoostingRegressor()
    gbdt.fit(X, y) # no warning
    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
    gbdt.fit(X, y) # raises warning

As already mentioned, the key to communication is the readability of the code
and good formatting can really be a plus. Notice that in the previous snippet
we:

>     - try to limit all lines to a maximum of 79 characters;

Reviewer: ```suggestion
    - try to limit all lines to a maximum of 79 characters to avoid horizontal
      scrollbars in the code snippets blocks rendered on the GitHub issue;
```

    - it should ideally only rely on a minimal dataset generated on-the-fly by
      running the code instead of relying on external data, if possible.

Use markdown formatting
-----------------------

To format code or text into its own distinct block, use triple backticks.
`Markdown
<https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax>`_
supports an optional language identifier to enable syntax highlighting in your
fenced code block. For example::

    ```python
    from sklearn.datasets import make_blobs

    n_samples = 100
    n_components = 3
    X, y = make_blobs(n_samples=n_samples, centers=n_components)
    ```

will render a python formatted snippet as follows

.. code-block:: python

    from sklearn.datasets import make_blobs

    n_samples = 100
    n_components = 3
    X, y = make_blobs(n_samples=n_samples, centers=n_components)

It is not necessary to create several blocks of code when submitting a bug
report. Remember other reviewers are going to copy-paste your code and having a
single cell will make their task easier.

> In the section `**Actual results**` of the `Issue template

Reviewer: ```suggestion
In the section named **Actual results** of the `Issue template
```

NumPy
-----

NumPy tools such as `numpy.random.randn
<https://numpy.org/doc/stable/reference/random/generated/numpy.random.randn.html>`_
and `numpy.random.randint
<https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html>`_
can be used to create dummy numeric data.

- regression

    Regressions take continuous numeric data as features and target.

    .. code-block:: python

        import numpy as np

        rng = np.random.RandomState(0)
        n_samples, n_features = 5, 5
        X = rng.randn(n_samples, n_features)
        y = rng.randn(n_samples)

A similar snippet can be used as synthetic data when testing scaling tools such
as :class:`sklearn.preprocessing.StandardScaler`.

- classification

    If the bug is not raised during when encoding a categorical variable, you can
    feed numeric data to a classifier. Just remember to ensure that the target
    is indeed an integer.

    .. code-block:: python

        import numpy as np

        rng = np.random.RandomState(0)
        n_samples, n_features = 5, 5
        X = rng.randn(n_samples, n_features)
        y = rng.randint(0, 2, n_samples)  # binary target with values in {0, 1}

>     If you need to test encoding, you may prefer to start from non-numeric data.
>     In such case you may use `numpy.random.choice

Reviewer: ```suggestion

    If the bug only happens with non-numeric class labels, you might want to
    generate a random target with `numpy.random.choice
```
